drivers/flash/flash_cadence_nand_ll.c

```diff
@@ -0,0 +1,1524 @@
+	ret = cdns_nand_wait_idle(base_address);
+	if (ret != 0) {
+		LOG_ERR("Wait for controller to be in idle state Failed");
+		return ret;
+	}
+
+	/*DMA Setting*/
+	sys_write32((F_BURST_SEL_SET(NF_TDEF_BURST_SEL)) | (BIT(F_OTE)),
+		    (base_address + NF_DMA_SETTING));
+
+	/*Pre fetch*/
+	sys_write32(((NF_FIFO_TRIGG_LVL_SET(PRE_FETCH_VALUE)) |
+		     (NF_DMA_PACKAGE_SIZE_SET(PRE_FETCH_VALUE))),
+		    (base_address + NF_PRE_FETCH));
+	/* Total bits in row addressing*/
+	params->total_bit_row = find_msb_set(((params->npages_per_block) - 1)) +
+				find_msb_set((params->nblocks_per_lun) - 1);
+
+	if (ret != 0) {
+		LOG_ERR("Failed to establish device access width!");
+		return -EINVAL;
+	}
+	/* Enable Global Interrupt for Nand*/
+#ifdef CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+	sys_set_bit((base_address + INTERRUPT_STATUS_REG), GINTR_ENABLE);
+#endif
+	return 0;
+}
+
+#if CONFIG_CDNS_NAND_CDMA_MODE
+/**
+ *
+ * This function performs Command descriptor structure prepareation.
+ *
+ * @param nf_mem determine which NF memory bank will be selected
+ * @param flash_ptr start ROW address in NF memory
+ * @param mem_ptr system memory pointer
+ * @param ctype Command type (read/write/erase)
+ * @param cmd_cnt counter for commands
+ * @param dma_sel select DMA engine (0 - slave DMA, 1 - master DMA)
+ * @param vol_id specify target volume ID
+ *
+ */
+void cdns_nand_cdma_prepare(char nf_mem, uint32_t flash_ptr, char *mem_ptr, uint16_t ctype,
+			    int32_t cmd_cnt, uint8_t dma_sel, uint8_t vol_id,
+			    struct cdns_cdma_command_descriptor *desc)
+{
+	struct cdns_cdma_command_descriptor *cdma_desc;
+
+	cdma_desc = desc;
+	/* set fields for one descriptor */
+	cdma_desc->flash_pointer = flash_ptr;
+	cdma_desc->bank_number = nf_mem;
+	cdma_desc->command_flags |= CDMA_CF_DMA_MASTER_SET(dma_sel) | F_CFLAGS_VOL_ID_SET(vol_id);
+	cdma_desc->memory_pointer = (uintptr_t)mem_ptr;
+	cdma_desc->status = 0;
+	cdma_desc->sync_flag_pointer = 0;
+	cdma_desc->sync_arguments = 0;
+	cdma_desc->ctrl_data_ptr = 0x40;
+	cdma_desc->command_type = ctype;
+	if (cmd_cnt > 1) {
+		cdma_desc->next_pointer = (uintptr_t)(desc + 1);
+		cdma_desc->command_flags |= CFLAGS_MPTRPC_SET | CFLAGS_MPTRPC_SET;
+		cdma_desc->command_flags |= CFLAGS_CONT_SET;
+	} else {
+		cdma_desc->next_pointer = 0;
+#ifdef CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+		cdma_desc->command_flags |= CDMA_CF_INT_SET;
+#endif
+	}
+}
+
+/**
+ * Check a command descriptor transfer complete status in the Cadence NAND controller.
+ *
+ * @param desc_ptr The pointer to the command descriptor structure.
+ * @param params The Cadence NAND parameters structure.
+ * @retval 0 on success or negative error value on failure.
+ */
+#ifdef CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+static int cdns_transfer_complete(struct cdns_cdma_command_descriptor *desc_ptr,
+				  struct cadence_nand_params *params)
+#else
+static int cdns_transfer_complete(struct cdns_cdma_command_descriptor *desc_ptr)
+#endif
+{
```

Comment:
Why just not define single cdsn_transfer_complete, with `struct cadence_nand_params *params` and do `ARG_UNUSED(params)` when not used?